src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -127,8 +127,8 @@ public static SourceText From(Stream stream, Encoding encoding = null, SourceHas
                 throw new InvalidDataException();
             }
 
-            var checksum = CalculateChecksum(stream, checksumAlgorithm);
-            return new StringText(text, encoding, checksum, checksumAlgorithm);
+            Debug.Assert(encoding != null);
+            return new StringText(text, encoding, checksumAlgorithm: checksumAlgorithm);
```

Comment:
Yes, it's possible that different source bytes will produce the same checksum, when the decoder is lossy. See the `FromStream_CheckSum_NoBOM` test as an example where `SourceText` is created with `Encoding.ASCII` but the actual bytes includes a value outside the ASCII range. However the checksum should be calculated with the decoded string rather than the original bytes since the decoded string is the text represented by the `SourceText`.